python/lsst/afw/geom/coordinates/coordinates.cc

```diff
@@ -0,0 +1,509 @@
+ * see <https://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include "pybind11/pybind11.h"
+#include "numpy/arrayobject.h"
+#include "ndarray/pybind11.h"
+
+#include "lsst/afw/geom/Point.h"
+#include "lsst/afw/geom/Extent.h"
+#include "lsst/afw/geom/CoordinateBase.h"
+#include "lsst/afw/geom/CoordinateExpr.h"
+
+namespace py = pybind11;
+using namespace pybind11::literals;
+
+namespace lsst {
+namespace afw {
+namespace geom {
+namespace {
+
+template <typename Derived, typename T, int N>
+using PyCoordinateBase = py::class_<CoordinateBase<Derived,T,N>>;
+
+template <int N>
+using PyCoordinateExpr = py::class_<CoordinateExpr<N>, CoordinateBase<CoordinateExpr<N>,bool,N>>;
+
+template <typename T, int N>
+using PyExtentBase = py::class_<ExtentBase<T,N>, CoordinateBase<Extent<T,N>,T,N>>;
+
+template <typename T, int N>
+using PyExtent = py::class_<Extent<T,N>, ExtentBase<T,N>>;
+
+template <typename T, int N>
+using PyPointBase = py::class_<PointBase<T,N>, CoordinateBase<Point<T,N>,T,N>>;
+
+template <typename T, int N>
+using PyPoint = py::class_<Point<T,N>, PointBase<T,N>>;
+
+
+template <typename Derived, typename T, int N>
+void declareCoordinateBase(py::module & mod, std::string const & suffix) {
+    static std::string const name = "CoordinateBase" + suffix;
+    PyCoordinateBase<Derived,T,N> cls(mod, name.c_str());
+
+    /* Operators */
+    cls.def("__getitem__", [](CoordinateBase<Derived,T,N> &c, int i) -> T {
+        if (i < 0 || i >= N) {
+            throw py::index_error{};
+        } else {
+            return c[i];
+        }
+    });
+    cls.def("__setitem__", [](CoordinateBase<Derived,T,N> &c, int i, T value) {
+        if (i < 0 || i >= N) {
+            throw py::index_error{};
+        } else {
+            c[i] = value;
+        }
+    });
+    cls.def("__len__", [](CoordinateBase<Derived,T,N> &c) -> int {
+        return N;
+    });
+}
+
+template <int N>
+void declareCoordinateExpr(py::module & mod, std::string const & suffix) {
+    static std::string const name = "CoordinateExpr" + suffix;
+    declareCoordinateBase<CoordinateExpr<N>,bool,N>(mod, name);
+
+    PyCoordinateExpr<N> clsCoordinateExpr(mod, name.c_str());
+
+    /* Constructors */
+    clsCoordinateExpr.def(py::init<bool>(), py::arg("val")=false);
+
+    /* Operators */
+    clsCoordinateExpr.def("and_", &CoordinateExpr<N>::and_);
+    clsCoordinateExpr.def("or_", &CoordinateExpr<N>::or_);
+    clsCoordinateExpr.def("not_", &CoordinateExpr<N>::not_);
+
+    mod.def("all", all<N>);
+    mod.def("any", any<N>);
+}
+
+template <typename T, int N>
+void declareExtentBase(py::module & mod, std::string const & suffix) {
+    declareCoordinateBase<Extent<T,N>,T,N>(mod, "Extent" + suffix);
+    static std::string const name = "ExtentBase" + suffix;
+
+    PyExtentBase<T,N> cls(mod, name.c_str());
+
+    /* Operators */
+    cls.def("eq", [](ExtentBase<T, N> const &self, Extent<T, N> other) { return self.eq(other); });
```

Comment:
A brief code comment explaining why these are not spelled `__eq__` etc. would be appreciated. Readers will wonder.